src/string_buffer.rs

```diff
@@ -80,22 +79,33 @@ impl StringBuffer {
     // Note that it is possible for this operation to take a long time in
     // pathological cases (lots of nodes, few line breaks).
     pub fn cur_offset(&self) -> usize {
-        unsafe {
-            let result = (&*self.last).cur_offset();
-            result.unwrap_or_else(|| {
-                println!("FIXME(#3) Unimplemented cur_offset across node boundaries");
-                0
-            })
-        }
+        // Try the last node first.
+        let result = unsafe {
+            (&*self.last).cur_offset()
+        };
+
+        // If the last node contains no newlines, try the nodes before.
+        let result = result.or_else(|| self.first.cur_offset());
+
+        // If there are no newlines at all, return the length of the buffer.
+        result.unwrap_or(self.len)
     }
 
     pub fn truncate(&mut self, new_len: usize) {
+        if new_len >= self.len {
+            return;
+        }
+
         let last = unsafe {
             &mut (*self.last).data
         };
-        // FIXME(#3) this will overflow if we need to truncate past the last node.
-        let last_len = last.len() - (self.len - new_len);
-        last.truncate(last_len);
+
+        // Check whether we need to truncate past the last node.
+        match last.len().checked_sub(self.len - new_len) {
```

Comment:
I would prefer to use `if last.len() > foo { ... }` rather than matching checked_sub - it makes it clearer what is going on and isn't as fragile wrt changing unsigned to signed ints.